# Lab 01 - Network traffic feature engineering

**下午第二節，約 60 分鐘。計算在這個 notebook 裡，判讀在 Grafana 上。**

Dashboard：<http://localhost:3000/d/aiops-lab01>（標題 `Lab 01 - Features and baselines`）

上午的結論是異常的定義：相對於一個明確 baseline 的偏離。數字大小本身不構成異常。
這節課把那句話拆成兩個可操作的決定。
第一個決定是 baseline，它決定同一筆讀數會不會被判為偏離。第二個決定是特徵，它決定那筆讀數存不存在。
兩個都選錯的時候，故障根本沒有進到判斷流程裡，連被誤判的機會都沒有。

**這節課結束時你要能回答的問題**：對 A 到 J 這十個事件，每一個要用哪個特徵、配哪一種 baseline 才看得見。
最後一個 cell 印出那張表的空白版本，由你填。

這份 notebook 不畫任何時序圖。圖在 dashboard 上，Grafana 負責畫，Python 負責算。
執行前先確認 exporter 與 Prometheus 都在跑；下一個 cell 的 `wk.bootstrap()` 會告訴你缺什麼。

| 時間 | 內容 |
| --- | --- |
| 00:00-00:08 | 單位契約與資料剖面。搞錯讀法不會報錯，只會小一個數量級 |
| 00:08-00:13 | 算完、推上 dashboard、設定 replay 速度 |
| 00:13-00:32 | 四個 baseline panel，用 Port dropdown 走一遍 |
| 00:32-00:40 | shape features 與退化的尺度 |
| 00:40-00:47 | Change point，以及四個名詞各自指什麼 |
| 00:47-01:00 | 三個必做的 GUI 練習，填表 |


In [1]:
import sys, pathlib

PROJECT_ROOT = pathlib.Path.cwd()
while PROJECT_ROOT != PROJECT_ROOT.parent and not (PROJECT_ROOT / "environment.yml").exists():
    PROJECT_ROOT = PROJECT_ROOT.parent
sys.path.insert(0, str(PROJECT_ROOT / "infra"))

import numpy as np
import pandas as pd
import aiopskit as wk

pd.set_option("display.width", 200)
pd.set_option("display.max_columns", 40)

wk.bootstrap()

project root  /Users/kevin/Library/CloudStorage/GoogleDrive-kevin880987@gmail.com/我的雲端硬碟/workshop/workbench/project/202608-aia-network-monitoring-aiops/aiops-anomaly-zero-to-hero
aiopskit      1.0.0

component             status    detail
------------------------------------------------------------------------------
Prometheus            up        http://localhost:9090
results exporter      up        http://localhost:8010/metrics
Grafana               up        http://localhost:3000  datasource=prometheus
  scrape prometheus     up        
  scrape rrd-exporter   down      
  scrape node-exporter  up        
  aiops_python_result             0 series in Prometheus

Prometheus is running but has no 'python-results-exporter' scrape job, so nothing you
  publish can reach Grafana. This happens when Prometheus was started from a
  package default config instead of the repository one. Stop it and restart with:
    prometheus --config.file=infra/prometheus/prometheus.macos.yml
  (use prometh

PosixPath('/Users/kevin/Library/CloudStorage/GoogleDrive-kevin880987@gmail.com/我的雲端硬碟/workshop/workbench/project/202608-aia-network-monitoring-aiops/aiops-anomaly-zero-to-hero')

## 1. 先把單位講清楚

把 `INOCTETS` 當成累積 counter 再對它做 `.diff()`，是這份資料上最容易犯、也最難察覺的一個錯。
沒有例外、沒有警告、圖也照樣畫得出來，只是圖上的數字全都小了一個數量級。
這一節從它開場，因為它示範了資料處理最危險的一種失敗：結果看起來完全正常。

`wk.load_telemetry()` 不猜。它印出它做了什麼決定，以及依據是什麼。
判準是 `monotonic_fraction`，它的定義與判讀方式在 Lab 00 第 9 節「counter 和 delta 的差別」。
loader 因此選了 `delta` 這條路：直接除以 cadence 得到每秒速率，不做差分。

下面第二段用 `semantics="counter"` 強迫誤讀一次，把代價量出來。正式流程不要這樣做。

In [2]:
telemetry, contract = wk.load_telemetry()
print(contract.describe())

CADENCE_S = contract.cadence_s
WINDOW = int(round(3600 / CADENCE_S))
print()
print(f"1h rolling window = {WINDOW} samples at {CADENCE_S:.0f}s cadence")

# 刻意誤讀成 counter，量出代價。正式流程不要這樣做。
wrong, wrong_contract = wk.load_telemetry(semantics="counter")

cost = pd.DataFrame({
    "delta_KBps": telemetry.groupby("port_id")["inoctets_ps"].mean() / 1000,
    "counter_KBps": wrong.groupby("port_id")["inoctets_ps"].mean() / 1000,
    "counter_zero_share": wrong.groupby("port_id")["inoctets_ps"].apply(lambda s: (s == 0).mean()),
})
cost["understated_by"] = cost["delta_KBps"] / cost["counter_KBps"]
print()
print("誤讀成 counter 的代價（每個 port 的平均 RX）")
print(cost.round(3).to_string())
print()
print(f"整體被 clip 成 0 的 INOCTETS 樣本比例  {(wrong['inoctets_ps'] == 0).mean():.4f}")
print(f"rows 差異                              {len(telemetry):,} -> {len(wrong):,}"
      f"  (每個 port 的第一筆做 diff 後為 NaN，被丟掉)")

rows            43,200 across 5 ports
span            2026-02-01 00:00:00  ->  2026-03-02 23:55:00
cadence         300 s per sample (5 min; 288 samples/day)
column reading  delta
monotonic frac  0.503 of consecutive octet steps are non-negative
note            reading chosen automatically: 50.3% of octet steps are non-negative, so the column is a per-interval total, not a counter

1h rolling window = 12 samples at 300s cadence



誤讀成 counter 的代價（每個 port 的平均 RX）
             delta_KBps  counter_KBps  counter_zero_share  understated_by
port_id                                                                  
port-id7427      82.101         5.434               0.496          15.109
port-id7428      17.238         1.180               0.501          14.612
port-id7429      62.208         4.827               0.496          12.888
port-id7430      31.002         2.063               0.495          15.029
port-id7431      38.371         2.828               0.498          13.569

整體被 clip 成 0 的 INOCTETS 樣本比例  0.4971
rows 差異                              43,200 -> 43,195  (每個 port 的第一筆做 diff 後為 NaN，被丟掉)


### 誤讀成 counter 的代價

拿 `port-id7429` 那一列來看。真實的平均接收速率是 62.2 KB/s，誤讀之後讀出來是 4.8 KB/s，
低估了將近 13 倍。原因在 `counter_zero_share` 那一欄：差分之後有 49.6% 的樣本是負數，
`clip(lower=0)` 把它們全部壓成零。
一半的樣本歸零，平均值當然崩掉，而且崩掉的方式是均勻的，所以曲線的形狀還在，只是被壓扁。
你看圖不會覺得哪裡不對。整體有 49.7% 的 `INOCTETS` 樣本會被壓成零。

差分還會吃掉每個 port 的第一筆（做 diff 之後是 NaN），43,200 列變成 43,195 列。
這五列本身不重要，重要的是它提醒你：差分會改變資料的長度、值域與分布，它有代價。

### 差分與否由欄位決定

Lab 00 裡你查過 `node_network_receive_bytes_total`，那個指標確實是 counter，
所以在 Grafana 裡對它用 `rate()` 是對的，不用 `rate()` 才是錯的。
同一台機器上，同一個 Prometheus 裡，兩個指標需要相反的處理。

「要不要差分」是欄位的屬性，由這個欄位被寫出來的方式決定，跟工具與課程慣例都無關。
唯一可靠的做法是先問這個欄位是怎麼被寫出來的，再決定怎麼讀。
`monotonic_fraction` 是一個便宜的檢查，它不能證明語意，但它可以在你想錯的時候攔下來。
RRD 匯出、SNMP polling 中介層、大部分廠商的 NMS 匯出檔，通常都是 delta；
Prometheus 的 `_total` 系列則幾乎都是 counter。拿到新資料先算這一個數字，成本是三行程式。

In [3]:
# 資料剖面。所有數字都從 telemetry 現算，沒有一個是寫死的常數。
PROFILE_COLS = ["traffic_bps", "packets_pps", "avg_pkt_bytes", "broadcast_ratio",
                "multicast_ratio", "tx_ratio", "error_rate", "discard_rate",
                "errors_pps", "unknown_proto_pps"]

print("=" * 96)
print("DATA PROFILE   data/synthetic/synthetic_rrd_metrics.csv")
print("=" * 96)
print(contract.describe())
print()
print(f"frame shape     {telemetry.shape[0]:,} rows x {telemetry.shape[1]} columns")
print(f"1h window       {WINDOW} samples;  1d window {int(round(86400 / CADENCE_S))} samples")
print()

inventory = (
    telemetry.groupby(["device_id", "port_id", "port_role"], observed=True)
    .agg(samples=("timestamp", "size"),
         first=("timestamp", "min"),
         last=("timestamp", "max"),
         median_MBps=("traffic_bps", lambda s: round(s.median() / 1e6, 4)),
         p95_MBps=("traffic_bps", lambda s: round(s.quantile(0.95) / 1e6, 4)))
    .reset_index()
)
print("port inventory   誰在哪台設備上，角色是什麼，流量級距差多少")
print(inventory.to_string(index=False))
print()

profile = pd.DataFrame({
    "nulls": telemetry[PROFILE_COLS].isna().sum(),
    "zero_frac": (telemetry[PROFILE_COLS] == 0).mean().round(3),
    "min": telemetry[PROFILE_COLS].min(),
    "median": telemetry[PROFILE_COLS].median(),
    "p95": telemetry[PROFILE_COLS].quantile(0.95),
    "max": telemetry[PROFILE_COLS].max(),
})
profile["p95_over_median"] = [
    round(p / m, 2) if m > 0 else np.nan for p, m in zip(profile["p95"], profile["median"])
]
profile["z_usable"] = np.where(profile["zero_frac"] > 0.5, "no", "yes")
print("per-feature profile   null 數、零讀數比例、四分位置")
print(profile.to_string(float_format=lambda v: f"{v:,.4g}", na_rep="-"))
print()

labelled = telemetry["event_label"] != "normal"
print(f"labelled samples   {int(labelled.sum()):,} / {len(telemetry):,} = {labelled.mean():.4f}"
      f"   （正常樣本 {int((~labelled).sum()):,} 筆）")
print()
event_inventory = (
    telemetry[labelled]
    .groupby(["event_id", "event_label"], observed=True)
    .agg(ports=("port_id", "nunique"), samples=("timestamp", "size"),
         start=("timestamp", "min"), end=("timestamp", "max"))
    .reset_index()
)
print("event inventory   直接從 event_label 這一欄數出來")
print(event_inventory.to_string(index=False))
print()

steps = (telemetry.sort_values(["port_id", "timestamp"])
         .groupby("port_id", observed=True)["timestamp"].diff().dt.total_seconds().dropna())
print("integrity")
print(f"  相鄰間隔的相異值        {sorted(steps.unique())} s over {len(steps):,} steps")
print(f"  重複的 (port, ts)       {int(telemetry.duplicated(['port_id', 'timestamp']).sum())}")
print(f"  全表 null 總數          {int(telemetry.isna().sum().sum())}")
print(f"  零讀數過半的欄位        {[c for c in PROFILE_COLS if (telemetry[c] == 0).mean() > 0.5]}")

DATA PROFILE   data/synthetic/synthetic_rrd_metrics.csv
rows            43,200 across 5 ports
span            2026-02-01 00:00:00  ->  2026-03-02 23:55:00
cadence         300 s per sample (5 min; 288 samples/day)
column reading  delta
monotonic frac  0.503 of consecutive octet steps are non-negative
note            reading chosen automatically: 50.3% of octet steps are non-negative, so the column is a per-interval total, not a counter

frame shape     43,200 rows x 34 columns
1h window       12 samples;  1d window 288 samples

port inventory   誰在哪台設備上，角色是什麼，流量級距差多少
 device_id     port_id      port_role  samples      first                last  median_MBps  p95_MBps
core-sw-01 port-id7429  server-uplink     8640 2026-02-01 2026-03-02 23:55:00       0.1451    0.3604
core-sw-01 port-id7430    office-vlan     8640 2026-02-01 2026-03-02 23:55:00       0.0425    0.1076
dist-sw-02 port-id7431 backup-storage     8640 2026-02-01 2026-03-02 23:55:00       0.0764    0.1926
edge-fw-01 port-id7427  

## 2. 讀這張剖面

上面那張表決定了後面每一個方法能不能成立，值得逐塊看過再往下。

`zero_frac` 那一欄先看。`error_rate` 與 `errors_pps` 的零讀數佔 97.6%，`unknown_proto_pps` 佔 99.0%，
`discard_rate` 佔 91.3%，而且前三者的 median 與 p95 同時是 0。
一個 12 筆的視窗裡若有十一筆是零，離散度就沒有東西可以估，`(x - center) / scale` 的分母會塌到下限，
商數反映的是分母有多小，跟偏離的程度無關。這四欄承載不了 z-score，
第 8 節把塌陷的比例量出來，並且給它們一個能用的替代做法。

剩下六欄可以用。`traffic_bps`、`packets_pps`、`avg_pkt_bytes`、`broadcast_ratio`、`multicast_ratio`、
`tx_ratio` 沒有任何 null，沒有任何零讀數，p95 對 median 的倍數落在 1.19 到 4.84 之間，
分布有肩膀也有尾巴。四種 baseline 與後面所有的帶號 z 分數都建在這六欄上。

port inventory 那一段給了一個限制。五個 port 的 median 從 `port-id7428` 的 0.028 MB/s
到 `port-id7429` 的 0.145 MB/s，差五倍，而且分屬 edge-fw-01、core-sw-01、dist-sw-02 三台設備。
拿原始 `traffic_bps` 做同儕比較，比的是不可比的東西，第 6.4 節會看到這個限制的實際後果。

cadence 是 300 秒，這件事有兩個後果。一小時的視窗是 12 筆樣本，一天 288 筆；
把 cadence 記成一分鐘的話，後面每一個 window 參數都會差六倍，而且同樣不會報錯。
另一個後果是解析度的上限：比 5 分鐘短的爆量在這份資料裡已經被 polling 平均掉了，它不存在，
任何以秒為單位的突發偵測在這份資料上做不出來。event inventory 裡最短的是事件 E，只有 8 筆樣本，
比一小時的視窗還短，Lab 02 要設「連續 N 筆超過門檻才確認」的時候，N 的上限就是被這種長度綁住的。

完整性沒有意外。43,195 個相鄰間隔全部是 300 秒，沒有缺樣本，沒有重複的 `(port_id, timestamp)`，
全表一個 null 都沒有。rolling window 的樣本數因此等於它的時間長度，不必先做重新取樣或補值。

最後是類別比例。43,200 筆裡有 286 筆帶事件標記，佔 0.66%，正常樣本 42,914 筆。
1% 的誤報率就會產生 429 則告警，比全部事件樣本還多。Lab 02 選評估指標的時候，這個比例是主要的約束。

## 3. 事件與 change calendar

真值有十個事件 A 到 J，剖面裡 `ports` 那一欄已經說了 G broadcast_storm 與 H multicast_flooding
同時打在五個 port 上，所以十個事件展開成 18 個 `(event, port)` 視窗。
事件層級的 recall 要用展開後的視窗算，不是用十個事件算。

`wk.split_events()` 拿 `change_calendar.csv` 去比對這 18 個視窗，分成兩堆：
16 個 incident 是偵測器必須抓到的，2 個 planned change 是偵測器必須安靜的。
兩堆在 telemetry 裡長得一模一樣，差別只在有沒有人事先登記過。

被判為 planned 的是 A business_traffic_growth 對上 CH-02，以及 C large_file_backup 對上 CH-01。
記住 C 是計畫內的，等一下在 Grafana 上它會是分數最高的那一段。

In [4]:
events = wk.load_events(telemetry)
calendar = wk.load_change_calendar()
incidents, planned = wk.split_events(events, calendar)

print(f"(event, port) windows  {len(events)}")
print(f"  incidents  必須抓到   {len(incidents)}")
print(f"  planned    必須安靜   {len(planned)}")
print()
print("planned changes:")
print(planned[["event_id", "event_type", "port_id", "start", "end", "change_id"]].to_string(index=False))
print()
print("change calendar:")
print(calendar[["change_id", "change_type", "scope", "start", "end", "owner", "expected_signal"]]
      .to_string(index=False))

(event, port) windows  18
  incidents  必須抓到   16
  planned    必須安靜   2

planned changes:
event_id              event_type     port_id               start                 end change_id
       A business_traffic_growth port-id7430 2026-02-04 09:00:00 2026-02-04 10:55:00     CH-02
       C       large_file_backup port-id7431 2026-02-08 01:00:00 2026-02-08 03:55:00     CH-01

change calendar:
change_id      change_type       scope               start                 end        owner                           expected_signal
    CH-02  business_change port-id7430 2026-02-04 08:50:00 2026-02-04 11:05:00     app-team              permanent level shift upward
    CH-01 scheduled_backup port-id7431 2026-02-08 00:50:00 2026-02-08 04:10:00    infra-ops sustained traffic surge on backup-storage
    CH-03      maintenance  core-sw-01 2026-02-16 02:00:00 2026-02-16 04:00:00  network-ops                             any deviation
    CH-04      maintenance  edge-fw-01 2026-02-21 03:00:00 2026-02-21 05

## 4. 四種 baseline 與 shape features

回傳值都是 `(center, scale)`，形狀一致，所以偏差分數永遠是 `(x - center) / scale`，
換 baseline 不用換分數的定義。這是 Lab 02 能把四個偵測器畫在同一軸上的前提。

`rolling_mean` 和 `rolling_robust` 都拿近期歷史當參考，差別在統計量：前者用 mean 與 std，
兩者都會被它要偵測的事件拉走；後者用 median 與 MAD，要超過半個視窗被污染才會動。
`seasonal` 換一個座標，按 weekday/weekend 加小時分組，比的是同一個季節位置。
`peer` 則完全不看歷史，只看同一瞬間的四個同儕，leave-one-out，對整批同時發生的事件是盲的。

前三種在單一 port 上算，peer 必須在五個 port 合併後才能算。順序不能反。

In [5]:
frames = []
for port_id, group in telemetry.groupby("port_id", observed=True):
    g = group.sort_values("timestamp").reset_index(drop=True)

    rc, rs = wk.baselines.rolling_mean(g["traffic_bps"], WINDOW)
    g["roll_center"], g["roll_lo"], g["roll_hi"] = rc, rc - 3 * rs, rc + 3 * rs
    g["z_rolling"] = wk.baselines.deviation_score(g["traffic_bps"], rc, rs)

    bc, bs = wk.baselines.rolling_robust(g["traffic_bps"], WINDOW)
    g["robust_center"], g["robust_lo"], g["robust_hi"] = bc, bc - 3 * bs, bc + 3 * bs
    g["z_robust"] = wk.baselines.deviation_score(g["traffic_bps"], bc, bs)

    sc, ss = wk.baselines.seasonal(g, "traffic_bps", by="daytype_hour")
    g["seasonal_center"], g["seasonal_lo"], g["seasonal_hi"] = sc, sc - 3 * ss, sc + 3 * ss
    g["z_seasonal"] = wk.baselines.deviation_score(g["traffic_bps"], sc, ss)

    frames.append(g)

feat = pd.concat(frames, ignore_index=True)

pc, ps = wk.baselines.peer(feat, "traffic_bps")
feat["peer_center"], feat["peer_lo"], feat["peer_hi"] = pc, pc - 3 * ps, pc + 3 * ps
feat["z_peer"] = wk.baselines.deviation_score(feat["traffic_bps"], pc, ps)

BASELINES = ["z_rolling", "z_robust", "z_seasonal", "z_peer"]
truth = wk.truth_mask(feat, events)

rows = []
for b in BASELINES:
    fired = feat[b].abs() > 4
    caught = sum(
        1 for e in events.itertuples()
        if fired[(feat["port_id"] == e.port_id)
                 & (feat["timestamp"] >= e.start)
                 & (feat["timestamp"] <= e.end)].any()
    )
    rows.append({
        "baseline": b.replace("z_", ""),
        "windows_caught_of_18": caught,
        "event_samples_over_4_%": round(100 * fired[truth].mean(), 1),
        "normal_samples_over_4_%": round(100 * fired[~truth].mean(), 2),
    })

print("同一條 traffic_bps，只換 baseline，|z| > 4")
print(pd.DataFrame(rows).to_string(index=False))
print()
print("沒有一種 baseline 抓得到 18 個視窗中的多數。問題在特徵，第 7 節那張表會指出來。")

同一條 traffic_bps，只換 baseline，|z| > 4
baseline  windows_caught_of_18  event_samples_over_4_%  normal_samples_over_4_%
 rolling                     0                     0.0                     0.00
  robust                     5                     8.0                     0.62
seasonal                     6                    38.8                     0.56
    peer                     4                    23.8                     0.96

沒有一種 baseline 抓得到 18 個視窗中的多數。問題在特徵，第 7 節那張表會指出來。


## 5. 推上 dashboard

到這裡為止的 cell 都執行過之後，`feat` 裡才有東西可以推。
`wk.publish()` 寫出哪些檔案、manifest 又約束了什麼，Lab 00 第 4 節「送上路」講過。

欄位名稱跟 dashboard 的 panel query 是綁死的。`roll_lo`、`robust_center`、`peer_hi` 這些名字改掉，
panel 就會變空白而且不會有任何錯誤訊息。要新增欄位可以，改名不行。

Lab 02 之後會 publish 兩個 lab 欄位的聯集，所以兩張 dashboard 都還會有資料。

`FOCUS` 那一行先維持 `None`，整個月都送出去。下面那兩段會告訴你什麼時候該改它。

In [6]:
PUBLISH_COLUMNS = [
    "traffic_bps",
    "roll_center", "roll_lo", "roll_hi",
    "robust_center", "robust_lo", "robust_hi",
    "seasonal_center", "seasonal_lo", "seasonal_hi",
    "peer_center", "peer_lo", "peer_hi",
    "error_rate", "discard_rate", "errors_pps",
    "avg_pkt_bytes", "broadcast_ratio", "multicast_ratio",
    "packets_pps",
]

# FOCUS = None 送整個月；給一組時間就只送那一段，讓 replay 變成短迴圈。
FOCUS = None
# FOCUS = ("2026-02-08 00:00", "2026-02-08 05:00")   # 事件 C large_file_backup

to_publish = feat
if FOCUS:
    lo, hi = pd.Timestamp(FOCUS[0]), pd.Timestamp(FOCUS[1])
    to_publish = feat[(feat["timestamp"] >= lo) & (feat["timestamp"] <= hi)]

manifest = wk.publish(to_publish, PUBLISH_COLUMNS, "lab01_baselines")

span_s = (to_publish["timestamp"].max() - to_publish["timestamp"].min()).total_seconds()
SCRAPE_S = 5
replay = pd.DataFrame([
    {
        "REPLAY_SPEED_X": sx,
        "full_pass_min": round(span_s / sx / 60, 1),
        "1_sim_hour_in_wall_s": round(3600 / sx, 1),
        "1_sample_in_wall_s": round(CADENCE_S / sx, 2),
        "scrapes_per_sample": round(CADENCE_S / sx / SCRAPE_S, 2),
    }
    for sx in (720, 360, 60)
])
print()
print(f"replay 換算（Prometheus scrape_interval = {SCRAPE_S}s，資料 cadence = {CADENCE_S:.0f}s）")
print(replay.to_string(index=False))
print()
print("scrapes_per_sample < 1 代表有資料點永遠不會被 scrape 到。")
print("dashboard  http://localhost:3000/d/aiops-lab01")

published 43,200 rows as 'lab01_baselines'
  archive   outputs/workshop/lab01_baselines.csv
  drop zone outputs/prometheus-dropzone/current_results.csv
  columns   traffic_bps, roll_center, roll_lo, roll_hi, robust_center, robust_lo, robust_hi, seasonal_center, seasonal_lo, seasonal_hi, peer_center, peer_lo, peer_hi, error_rate, discard_rate, errors_pps, avg_pkt_bytes, broadcast_ratio, multicast_ratio, packets_pps
  query     aiops_python_result{column="traffic_bps"}

replay 換算（Prometheus scrape_interval = 5s，資料 cadence = 300s）
 REPLAY_SPEED_X  full_pass_min  1_sim_hour_in_wall_s  1_sample_in_wall_s  scrapes_per_sample
            720           60.0                   5.0                0.42                0.08
            360          120.0                  10.0                0.83                0.17
             60          719.9                  60.0                5.00                1.00

scrapes_per_sample < 1 代表有資料點永遠不會被 scrape 到。
dashboard  http://localhost:3000/d/aiops-lab01


### 一個會讓你卡住的陷阱

第二個 cell 的 `wk.bootstrap()` 會檢查整條路徑。如果你看到這段輸出：

```text
Prometheus is running but has no 'python-results-exporter' scrape job, so nothing you
  publish can reach Grafana. This happens when Prometheus was started from a
  package default config instead of the repository one. Stop it and restart with:
    prometheus --config.file=infra/prometheus/prometheus.macos.yml
```

成因與各平台設定檔的名字在 Lab 00 第 1 節「那個會讓你浪費半小時的坑」。

### replay 速度決定你看得到什麼

exporter 把一個月的資料按牆上時間播放。`REPLAY_SPEED_X` 是每一秒真實時間推進多少秒資料時間。
上面那張換算表的 `scrapes_per_sample` 是重點。
在 720x 之下每筆樣本只被 scrape 0.08 次，也就是十二筆資料裡只有一筆會進到 Prometheus。
把一個月的資料壓進一堂課，必然要付這個代價，設定本身沒有錯。
事件 G 只有 12 筆樣本，在 720x 之下總共佔 5 秒牆上時間，剛好是一個 scrape interval，
所以它有可能整個掉在兩次 scrape 之間，一個點都不留。

**Grafana 的時間選擇器幫不上忙。** 下面給的所有時間戳（`2026-02-08 01:00` 之類）都是資料時間，
牆上時間要另外換算。你在時間選擇器裡輸入 2026 年 2 月，畫面一定是空的，因為那個牆上時刻沒有任何資料點。
Panel 用的是相對區間 `now-30m`，你看到的永遠是「現在正在播的那一段」。

有三種辦法處理這件事，各自看得到不同的東西，選一個，不要混著用：

1. **維持 720x，看整個月的形狀。** 一輪 60 分鐘。用 Lab 00 dashboard 的 `Replay progress`
   判斷播到哪裡了；那塊 tile 的 unit 是 percentunit，所以下面導航表寫的 0.113 在畫面上是 11.3%。
   下面那張導航表給每個事件的 progress 值與出現時間。
   這個模式適合看日夜週期、看 seasonal baseline 的輪廓，不適合看單一事件的細節。
2. **改成 60x，讓每筆樣本剛好被 scrape 一次。** 這是唯一能在 Grafana 上重現 notebook 計算的模式，
   代價是播完一輪要 12 小時，事件 C 要等 169 分鐘才出現。整輪播放在課堂上不可行。
3. **改成 60x，同時只 publish 事件附近那一段。** 這是要細看某個事件時該用的做法。
   回到上面那個 publish cell，把 `FOCUS` 那一行取消註解：

   ```python
   FOCUS = ("2026-02-08 00:00", "2026-02-08 05:00")   # 事件 C large_file_backup
   ```

   重跑那個 cell，drop zone 只剩 305 列，replay 變成 5 分鐘的短迴圈，每筆樣本都被 scrape 到，
   事件 C 的三個小時佔滿整個畫面。看完記得改回 `FOCUS = None` 再重跑。

改速度要重起 exporter：`REPLAY_SPEED_X=60 python infra/python_results_exporter.py`。

720x 之下的導航表：

| 事件 | 資料時間起點 | 樣本數 | replay progress | 開播後第幾分鐘 | 在畫面上停留 |
| --- | --- | --- | --- | --- | --- |
| A business_traffic_growth | 2026-02-04 09:00 | 24 | 0.113 | 6.8 | 10.0 秒 |
| B small_packet_scan | 2026-02-06 13:10 | 10 | 0.185 | 11.1 | 4.2 秒 |
| C large_file_backup | 2026-02-08 01:00 | 36 | 0.235 | 14.1 | 15.0 秒 |
| D queue_congestion | 2026-02-11 15:20 | 14 | 0.355 | 21.3 | 5.8 秒 |
| E link_quality_issue | 2026-02-14 10:35 | 8 | 0.448 | 26.9 | 3.3 秒 |
| F load_sensitive_link_issue | 2026-02-17 16:00 | 18 | 0.556 | 33.3 | 7.5 秒 |
| G broadcast_storm | 2026-02-19 11:30 | 12 | 0.616 | 37.0 | 5.0 秒 |
| H multicast_flooding | 2026-02-22 20:05 | 18 | 0.728 | 43.7 | 7.5 秒 |
| I abnormal_device_sender | 2026-02-25 09:45 | 16 | 0.814 | 48.8 | 6.7 秒 |
| J unknown_protocol_scan | 2026-02-27 14:15 | 10 | 0.887 | 53.2 | 4.2 秒 |

## 6. 四個 baseline，同一條線

打開 <http://localhost:3000/d/aiops-lab01>。最上面是一個叫 **Compared with what?** 的文字 panel，
它是這一頁的題目；旁邊的 **Port** dropdown 是這一頁唯一的控制項。
往下四個 timeseries panel 畫的是**同一條** `traffic_bps`，差別只在陰影帶是誰算出來的。

**讀 panel 之前先在 Port dropdown 裡挑一個 port，不要停在預設的 `All`。**
選 `All` 不會出錯，五個 port 的線和五組陰影帶會一起畫上去，但這一頁比較的是同一條線在四種 baseline 底下的判定，
五條線疊在一起就沒有那個「同一條」可言了。一次只看一個 port，是這四張圖能並排的前提。
把 Port 切到 `port-id7431`（backup-storage，dist-sw-02）。

建議用上面第 3 種模式做這一段：`FOCUS = ("2026-02-08 00:00", "2026-02-08 05:00")`，`REPLAY_SPEED_X=60`。
五分鐘一輪，看不清楚就等下一輪。

### 6.1 Recent history: 1h rolling mean, +/- 3 sigma

事件 C `large_file_backup` 打在 port-id7431，資料時間 2026-02-08 01:00 到 03:55，共 36 筆樣本。
視窗是 12 筆。事件比視窗長三倍，這是接下來所有現象的根源。

事件前一小時的流量大約 0.011 MB/s，事件期間平均 0.060 MB/s，尖峰 0.075 MB/s，差了五倍以上。
看 panel 上的陰影帶，你會看到兩件事同時發生。

第一筆事件樣本進來的時候（01:00），rolling mean 從 0.011 動到 0.014，動得不多；
但 rolling 標準差從 0.004 跳到 0.012，帶寬從 0.024 MB/s 撐開到 0.070 MB/s。
一筆樣本就把容忍區間撐寬了將近三倍。這一筆的分數是 3.16，不到門檻 4。
如果標準差沒有跟著動，同一筆的分數會是 10.0，遠遠過關。
**是那筆異常樣本自己把自己的容忍區間撐開的。**

接下來的兩小時，mean 一路被拉上去。01:30 時中心已經到 0.039，02:00 到 0.059，
也就是說 baseline 追上了事件，而事件還在進行中。從 02:00 之後分數在零附近上下晃，
panel 上看起來完全正常。整整 36 筆樣本裡，`|z| > 4` 的樣本數是 **0**。
這個事件在這個 panel 上完全沒有被記錄過。

還有一個現象值得看：04:00 流量掉回 0.015 MB/s 的那一刻，分數是 -2.91。
事件結束跟事件開始一樣是偏離。用單邊（只看高於 baseline）的偵測器會漏掉這一半。

### 6.2 Recent history, robust: 1h rolling median, +/- 3 MAD

同一段時間、同一條線，換成 median 與 MAD。

前五筆事件樣本（01:00 到 01:20），中位數紋風不動，還停在 0.011，帶寬也還是 0.023 MB/s。
分數是 10.34、10.08、14.26、10.84、15.10，全部過門檻。

第六筆（01:25）中位數突然跳到 0.031，分數塌到 1.29，之後就再也沒有回來。
這個時點可以事先算出來：視窗 12 筆，第六筆進來之後事件樣本剛好占滿一半，
中位數失去支撐點。中位數要超過半個視窗被污染才會動，這句話的另一面是**一旦超過就一定會動**。
36 筆裡有 5 筆過門檻，只有前 25 分鐘。

兩個 baseline 最後都會吸收比視窗長的事件，差別在過程：平均數從第一筆就開始投降，
而且是連續的、每一筆都退一點；中位數在前五筆完全不退，然後一次全部投降。
偵測器要的是事件剛開始那幾筆的分數，所以中位數這種「撐住再垮」的形狀，比平均數那種「一路軟化」的形狀有用得多。

帶寬本身反而是相反的：rolling 帶寬的最大值是 0.153 MB/s，robust 帶寬的最大值是 0.193 MB/s，
robust 那條後來張得更開。如果你只盯著陰影帶的寬度看，會得到錯誤的結論。要看的是線有沒有跑出帶外，
以及跑出去的那幾筆是不是在事件的開頭。

### 6.3 Same seasonal position: this hour, on days like today

同樣是事件 C，這個 panel 的 36 筆樣本**全部** `|z| > 4`，最小 9.9，中位數 12.4，最大 16.1。
從頭到尾沒有衰減。

原因很單純：這個 baseline 完全不看剛剛發生了什麼。它把歷史按「平日或假日」加「小時」分組，
拿同一格的中位數當中心。凌晨一點在 port-id7431 上的正常值是 0.011 MB/s，
不管過去一小時發生過什麼事，這個判斷都不會改變。事件拉不動它，因為事件不在它的輸入裡。

把 Port 切到 `port-id7430`（office-vlan），看這個 baseline 的中心線在一天之內的形狀。
平日各小時的中位數是這樣：

```text
00h 0.0106   06h 0.0286   12h 0.1020   18h 0.0726
03h 0.0118   09h 0.0804   15h 0.0987   21h 0.0303
```

早上七點到十點之間中心線爬升將近一倍，這是**唯一**一個知道「早上流量本來就該漲」的 baseline。
其他三個都不知道，它們只知道現在跟剛剛不一樣。

這裡要誠實說清楚：在一個正常的平日早晨，rolling 與 robust 也不會誤報。
2026-02-11 早上六點到十一點的 `port-id7430`，rolling 的 `|z|` 最高 2.42，robust 最高 2.12，
seasonal 最高 2.80，三個都離門檻 4 有一段距離。爬升是緩慢的，以 12 筆樣本的視窗來看，每一步都在容忍範圍內。
差別要在真正快速而且持續的位移上才會出現。看事件 A（`port-id7430`，2026-02-04 09:00 到 10:55，24 筆）：

| baseline | 事件期間的表現 |
| --- | --- |
| rolling mean | 峰值 3.00，從未過門檻 |
| robust | 前兩筆 4.73 與 7.00 過門檻，第六筆起塌到 1.36，之後最高只回到 1.60 |
| seasonal | 24 筆全部 ≥ 4.4，峰值 10.65 |

一個註記留給課後：`wk.baselines.seasonal()` 有一個 `train_mask` 參數，
可以把已知事件視窗排除在 profile 之外。這份 lab 沒有用它，跟 Lab 02 保持一致。
不排除的代價是 profile 吸收了一部分事件，等於用「含有上個月那次事故的歷史」在定義正常。
自己改一次跑跑看，那是這門課裡最容易被忽略的一種資料汙染。

### 6.4 Peer group: this port against its siblings, same instant

把 Port 切到 `port-id7428`，事件 G `broadcast_storm`，資料時間 2026-02-19 11:30 到 12:25，12 筆樣本。
在 720x 之下這是開播後第 37 分鐘、畫面上 5 秒；要細看請用 `FOCUS = ("2026-02-19 11:00", "2026-02-19 13:00")`。

**先確認它是同時發生的。** 在 Port dropdown 之間切換 7427、7428、7429、7430、7431，
每一個 port 的事件視窗都是 11:30 到 12:25，一筆不差。這五個 port 分屬三台設備，卻在同一個 5 分鐘刻度上一起動。
L2 廣播風暴本來就是這樣：它不挑 port。

現在看數字。`broadcast_ratio` 在事件期間五個 port 分別穩定在
0.0469、0.0458、0.0371、0.0410、0.0420，平時的中位數大約 0.003，也就是漲了十幾倍。
拿這條 `broadcast_ratio` 去比兩種 baseline：

| port | robust (1h median/MAD) | peer (leave-one-out) |
| --- | --- | --- |
| port-id7427 | 50.0（截斷） | 1.5 |
| port-id7428 | 50.0（截斷） | 1.2 |
| port-id7429 | 50.0（截斷） | 1.9 |
| port-id7430 | 50.0（截斷） | 0.8 |
| port-id7431 | 50.0（截斷） | 0.3 |

robust baseline 拿現在跟自己的過去比，看到十幾倍的跳升，分數直接頂到上限。
peer baseline 拿現在跟同一瞬間的其他四個 port 比，而其他四個 port 也漲了同樣的倍數，
所以「相對於同儕」這件事完全沒有改變。**這個 baseline 對事件 G 是全盲的。**

事件 H `multicast_flooding` 是一樣的形狀，`multicast_ratio` 的 robust 分數五個 port 全部 50.0，
peer 分數是 4.4、1.5、0.4、0.3、1.8。

盲和有用是同一個性質的兩面。假設今晚有人改了全公司的備份策略，五個 port 一起漲。
rolling、robust、seasonal 三個 baseline 會同時對五個 port 送出告警，值班的人一次收到五封通知，
而且五封講的是同一件事。peer baseline 一封都不會發。
所以 peer 是唯一一個能扛住全網位移的 baseline，也是唯一一個看不見全網故障的 baseline。
它從頭到尾回答的是同一個問題：這個 port 有沒有比它的同儕異常。

還有一個限制要講，剖面裡已經先看到了。這五個 port 的角色與流量級距差很多，
`port-id7429` 的正常中位數是 0.145 MB/s，`port-id7428` 只有 0.028 MB/s，差五倍。
事件 A 的 peer 分數只有 0.7，事件 I 只有 2.8，兩個都是單一 port 的真實異常，peer 卻看不見。
同儕比較要有意義，同儕必須真的可比。實務上這代表你得先按角色分群，或者比 ratio 而不是比絕對量。

## 7. 特徵決定看不看得見

捲到 **Shape features: mean packet size, broadcast and multicast share** 這個 panel。
上面四個 panel 換了四種 baseline，都在看 `traffic_bps` 這一條線。現在換一條線來看。

### 事件 B：流量沒動，封包全變了

事件 B `small_packet_scan`，port-id7428，2026-02-06 13:10 到 13:55，10 筆樣本。

兩欄都取事件開始前十二筆樣本的中位數當作事件前的水準。這裡不能拿整月中位數來比，
事件落在下午的流量高峰上，整月中位數只有 34 pps，用它會把事件本身的倍率灌水到二十倍。

| | 事件前 | 事件期間 |
| --- | --- | --- |
| `packets_pps` | 82 | 667 |
| `avg_pkt_bytes` | 764 bytes | 127 bytes |
| `traffic_bps` | 65.6 kB/s | 80.4 kB/s |
| `traffic_bps` 的 robust 分數 | | 起始 1.37，全程峰值 2.75 |
| `avg_pkt_bytes` 的 robust 分數 | | 起始 **-14.82** |

封包數變成八倍，平均封包大小掉到六分之一。八乘上六分之一約等於一點三，
所以位元組流量只上升兩成多，`traffic_bps` 的分數全程進不了門檻。
四種 baseline 都救不了它，因為要看的東西根本不在這條線裡。

流量是封包數乘上封包大小。任何一種「拿其中一項換另一項」的故障，在乘積上都是隱形的。
掃描、探測、小封包 flood 全屬於這一類。要看見它們，你得去看那兩個因子本身。

**注意符號是負的。** `avg_pkt_bytes` 在事件期間是往**下**掉。
一個只判斷「高於 baseline」的偵測器會完整地錯過事件 B。方向要跟門檻一起設計。

下面把 18 個視窗、8 個特徵的帶號峰值 z 一次印出來，baseline 固定用 1h robust。

In [7]:
# traffic_bps 的 1h robust 分數第 4 節已經算過，這裡只補其餘七欄。
SHAPE_FEATURES = ["packets_pps", "avg_pkt_bytes", "broadcast_ratio",
                  "multicast_ratio", "error_rate", "discard_rate", "tx_ratio"]
FEATURES = ["traffic_bps"] + SHAPE_FEATURES

scored = []
for port_id, group in telemetry.groupby("port_id", observed=True):
    g = group.sort_values("timestamp").reset_index(drop=True)
    for col in SHAPE_FEATURES:
        cen, sca = wk.baselines.rolling_robust(g[col], WINDOW)
        g["z_" + col] = wk.baselines.deviation_score(g[col], cen, sca)
    scored.append(g)
# 兩個迴圈的分組與排序相同，feat 的列序與 scored 一致。
scored = pd.concat(scored, ignore_index=True)
scored["z_traffic_bps"] = feat["z_robust"].to_numpy()


# 視窗內離 0 最遠的那一點，保留符號。
def signed_peak(series):
    return series.loc[series.abs().idxmax()]


table = []
for e in events.itertuples():
    window = scored[(scored["port_id"] == e.port_id)
                    & (scored["timestamp"] >= e.start)
                    & (scored["timestamp"] <= e.end)]
    row = {"ev": e.event_id, "type": e.event_type, "port": e.port_id[-4:]}
    row.update({col: round(signed_peak(window["z_" + col]), 1) for col in FEATURES})
    table.append(row)

print("每個事件視窗內的帶號峰值 z（1h robust baseline，50.0 = 截斷，讀成 off the scale）")
print(pd.DataFrame(table).to_string(index=False))

每個事件視窗內的帶號峰值 z（1h robust baseline，50.0 = 截斷，讀成 off the scale）
ev                      type port  traffic_bps  packets_pps  avg_pkt_bytes  broadcast_ratio  multicast_ratio  error_rate  discard_rate  tx_ratio
 A   business_traffic_growth 7430          7.0          4.5           -2.2            -50.0            -50.0        13.1           0.8       2.3
 B         small_packet_scan 7428          2.8         50.0          -14.8            -50.0            -50.0         0.0           0.0       4.7
 C         large_file_backup 7431         15.1          3.1           14.2            -17.2            -26.3         0.0           9.2       2.5
 D          queue_congestion 7427         40.4         33.1            3.5            -50.0            -50.0         0.0          50.0      -3.3
 E        link_quality_issue 7429          3.6          2.3            3.4             -2.1              2.9        50.0           1.7       2.9
 F load_sensitive_link_issue 7427         14.6         -4.8         

讀這張表的方式是逐列問一句話：這一列有哪幾欄的絕對值超過 4。
如果只有 `traffic_bps` 那一欄過不了 4，代表流量本身看不見這個故障。
符號也要看，`avg_pkt_bytes` 在 B small_packet_scan 那一列是負的。

### 事件 G：同一件事，換個角度

事件 G 在 `traffic_bps` 上，五個 port 的 robust 分數落在 1.5 到 3.7 之間，沒有一個過 4。
第 6.4 節看到，同一個事件在 `broadcast_ratio` 上五個 port 全部頂到上限。

所以事件 G 有兩層盲點，要分開講清楚。第一層是特徵：看 `traffic_bps` 的話，**沒有任何 baseline** 看得見它。
第二層是 baseline：改看 `broadcast_ratio` 之後，robust 看得一清二楚，peer 還是全盲。
特徵先，baseline 後。特徵選錯的時候，調 baseline 是在浪費時間。

那張表還有一個容易誤讀的地方。`broadcast_ratio` 在事件 A、B、D、J 上都是 -50.0，事件 I 上是 -47.9。
它不是 broadcast storm 的專屬指紋。只要總封包數大幅變動，比值就會被牽動，
而它的 1h MAD 很小，任何牽動都會被放大。特徵會響只代表值得看一眼，離「就是這個故障」還有距離。
分辨要靠多個特徵的組合，以及符號：事件 G 的 `broadcast_ratio` 是 +50，事件 A、B、D 是 -50。

## 8. 退化的尺度

先看 **Quality features: error and discard rate** 這個 panel，把 Port 切到 `port-id7429`。
兩條線幾乎整段貼在零上，偶爾有一根刺，然後又回到零。
這種形狀在 dashboard 上很容易被當成「沒事」而跳過，但它其實是這節課最重要的一張圖，
因為它讓後面所有的統計方法失效。剖面裡 `zero_frac` 那一欄已經預告了這件事，下面把代價量出來。

`wk.baselines.degenerate_fraction()` 量的是滾動 scale 塌到下限的樣本比例。
塌到下限代表那段視窗裡幾乎沒有離散度可言。
最後幾行順便量一下 Lab 02 那條固定門檻 `errors_pps > 0.05` 的分離度。

In [8]:
DEGEN_COLS = ["traffic_bps", "error_rate", "discard_rate", "broadcast_ratio",
              "multicast_ratio", "avg_pkt_bytes", "unknown_proto_pps"]

per_port = {}
for port_id, group in telemetry.groupby("port_id", observed=True):
    g = group.sort_values("timestamp").reset_index(drop=True)
    per_port[port_id] = {
        col: wk.baselines.degenerate_fraction(
            wk.baselines.rolling_robust(g[col], WINDOW)[1], g[col])
        for col in DEGEN_COLS
    }

degen = pd.DataFrame(per_port).round(3)
degen["mean"] = degen.mean(axis=1).round(3)
degen["zero_share"] = [round((telemetry[c] == 0).mean(), 3) for c in DEGEN_COLS]

print("degenerate scale fraction：滾動 scale 塌到下限的樣本比例")
print(degen.to_string())
print()

over = telemetry["errors_pps"] > 0.05
print(f"errors_pps > 0.05 的樣本數        {int(over.sum())} / {len(telemetry):,}")
print(f"其中落在已標記事件視窗內的比例    {wk.truth_mask(telemetry, events)[over].mean():.3f}")
print(f"非零但低於門檻的背景值            {sorted(set(telemetry.loc[(telemetry['errors_pps'] > 0) & ~over, 'errors_pps'].round(4)))}")

degenerate scale fraction：滾動 scale 塌到下限的樣本比例
                   port-id7427  port-id7428  port-id7429  port-id7430  port-id7431   mean  zero_share
traffic_bps              0.199        0.204        0.209        0.208        0.204  0.205       0.000
error_rate               0.998        1.000        0.998        1.000        1.000  0.999       0.976
discard_rate             0.992        0.995        0.995        0.995        0.995  0.994       0.913
broadcast_ratio          0.001        0.001        0.001        0.001        0.001  0.001       0.000
multicast_ratio          0.001        0.001        0.001        0.001        0.001  0.001       0.000
avg_pkt_bytes            0.001        0.001        0.001        0.001        0.001  0.001       0.000
unknown_proto_pps        1.000        0.998        1.000        1.000        1.000  1.000       0.990

errors_pps > 0.05 的樣本數        20 / 43,200


其中落在已標記事件視窗內的比例    1.000
非零但低於門檻的背景值            [0.0033, 0.0067, 0.01]


拿 `port-id7429`（server-uplink，事件 E 發生的那個 port）那一欄來看。
`unknown_proto_pps` 是 1.000，字面意思就是**沒有任何一個視窗**有可用的尺度估計。
`error_rate` 是 0.998，而且它有 97.6% 的樣本讀數是精確的零。
一條健康的線路本來就不該有錯誤，所以這個結果是這個指標的本性，資料本身沒有問題。

問題在於：當分母幾乎是零，`(x - center) / scale` 這個式子不再是一個量測。
任何非零讀數除下去都會得到天文數字，而那個數字跟「偏離的程度」沒有關係，它只反映了分母有多小。
`deviation_score()` 把結果截斷在 `MAX_SCORE = 50`，所以你在表上看到 50.0 要讀成 off the scale，
不能讀成「五十個標準差」。截斷是為了讓你在排序告警的時候，不會被一個除以零的產物排到最前面。

**這是關於這個指標的陳述。** z-score 沒有壞掉，它算得完全正確；
是這個指標不提供 z-score 需要的東西。換一個更聰明的偵測器不會改變任何事，因為問題在輸入端。
近乎常數的指標要的是固定門檻。

### 固定門檻在這裡能用的理由

Lab 02 的第三條告警規則是 `errors_pps > 0.05`，用原始計數而不是分數。上面那三行已經把數字印出來。

站得住腳的條件有兩個。這個指標有一個由物理決定的自然參考點，零錯誤；訊號與背景之間又隔著兩個數量級的空白。
兩個條件同時成立，這條門檻才成立。
換一個沒有自然參考點的指標，同樣的做法會立刻變成一場門檻的猜謎。

### 給這條規則一個正確的名字

早上列了七種 baseline 來源，這個 lab 到這裡用掉三種：recent history（6.1 與 6.2）、
same seasonal position（6.3）、peer group（6.4）。`errors_pps > 0.05` 是另外兩種疊在一起。
「一個介面本來就不該有錯誤」是 physical limit，這句話由乙太網路本身給定，跟這份資料學到了什麼無關；
「門檻放在 0.05」則是 expert knowledge，由看過夠多條線路的人決定放哪裡。
這兩種來源有一個共同性質值得記住：它們不需要歷史。前面三種 baseline 每一種都要先攢滿一個月的資料才算得出來，
而你接手一台全新設備的那個早上，手上能立刻用的就只有 physical limit 跟 expert knowledge 這兩種。

### 剩下兩種：SLO 與 upstream/downstream

這兩種在這份資料上沒有現成的形式，位置還是說得出來，名字也必須先講在這裡。

SLO baseline 拿對外承諾當比較基準，數字來自合約。
把 `port-id7427` 這條 wan-primary 的承諾寫成「95 百分位不超過 0.25 MB/s」，偏離的定義就從「離平常多遠」
換成「超出承諾多久」。剖面那張 port inventory 已經給了這五個 port 目前的 95 百分位，
從 port-id7427 到 port-id7431 依序是 0.2493、0.0694、0.3604、0.1076、0.1926 MB/s。
承諾線畫在哪裡要先跟業務談，談完才輪到統計。

upstream/downstream 比較則是把同一批流量在鏈路不同位置的讀數對起來看，兩端的差額本身就是一個 baseline。
這五個 port 分屬三台設備，
可是這份資料只有 `device_id`，沒有任何一欄說誰接在誰後面，所以「上游看得到、下游少了一截」這種比法在這裡做不出來。
拿到真實網路的第一件事是問拓樸表在哪，那一欄的價值不輸給任何一種統計量。

先記住這個伏筆：Lab 02 的偵測器用 `traffic_bps`、`broadcast_ratio`、`multicast_ratio`、`avg_pkt_bytes`
這四個特徵做 robust z，16 個 incident 抓到 15 個，唯一漏掉的就是 port-id7429 上的事件 E。
漏掉的原因在特徵：能看見 E 的那個特徵，本質上不能用 z-score 去看，調門檻與視窗長度都改變不了這件事。

## 9. Change point：水位有沒有整段移動

前面每一個分數回答的都是同一個問題：這一筆讀數離它該在的位置有多遠。
還有一種問法形狀完全不同。它看的是一段跟另一段之間的落差，也就是這條線的水位有沒有在某個時刻整個移上去。

最省事的做法是雙移動平均。短視窗與長視窗各算一次平均，取兩者的相對落差：

```text
(fast - slow) / |slow|
```

水位平穩的時候快慢兩條疊在一起，落差在零附近。水位一移動，快的那條先跟上新水位，慢的那條還留在舊水位，
落差因此撐開，直到長視窗也走完新的一段為止。`wk.detect.score_dual_ma(values, short, long)` 就是這個式子。
它的代價是延遲：慢的那條必須先停止追蹤舊水位，落差才顯示得出來。

下面在 `port-id7430` 的 `traffic_bps` 上算一次，短視窗 12 筆（一小時），長視窗 72 筆（六小時），
再對十個事件視窗各取一次帶號峰值。`hits_this_port` 那一欄標的是這個事件到底有沒有打在這個 port 上。
最後三行去查 CH-02 在 change calendar 上宣告的 `permanent level shift upward` 到底有沒有留下來。

In [9]:
CP_PORT = "port-id7430"
SHORT, LONG = WINDOW, 6 * WINDOW

cp = telemetry[telemetry["port_id"] == CP_PORT].sort_values("timestamp").reset_index(drop=True)
dma = wk.detect.score_dual_ma(cp["traffic_bps"], short=SHORT, long=LONG)

rows = []
for e in events.drop_duplicates("event_id").itertuples():
    seg = dma[(cp["timestamp"] >= e.start) & (cp["timestamp"] <= e.end)].dropna()
    rows.append({
        "ev": e.event_id,
        "type": e.event_type,
        "hits_this_port": bool(((events["event_id"] == e.event_id)
                                & (events["port_id"] == CP_PORT)).any()),
        "peak_dual_ma": round(float(seg.loc[seg.abs().idxmax()]), 2),
    })

divergence = pd.DataFrame(rows).sort_values("peak_dual_ma", key=lambda s: s.abs(), ascending=False)

print(f"dual MA on {CP_PORT}.traffic_bps    short={SHORT} long={LONG} samples "
      f"({SHORT * CADENCE_S / 3600:.0f}h vs {LONG * CADENCE_S / 3600:.0f}h)")
print(divergence.to_string(index=False))
print()
print(f"同一條線上 |dual MA| 的日常水準  median {dma.abs().median():.2f}   p90 {dma.abs().quantile(0.9):.2f}")
print()

# CH-02 宣告 permanent level shift upward。比對事件前後同一個時段的水位。
slot = cp[(cp["timestamp"].dt.hour >= 9) & (cp["timestamp"].dt.hour < 11)
          & (cp["timestamp"].dt.dayofweek < 5)]
before = slot.loc[slot["timestamp"] < "2026-02-04", "traffic_bps"]
after = slot.loc[slot["timestamp"] > "2026-02-05", "traffic_bps"]
print("CH-02 宣告 permanent level shift upward；平日 09:00-11:00 的實測中位數")
print(f"  事件之前  {before.median() / 1e6:.4f} MB/s   n={len(before)}")
print(f"  事件之後  {after.median() / 1e6:.4f} MB/s   n={len(after)}")

dual MA on port-id7430.traffic_bps    short=12 long=72 samples (1h vs 6h)
ev                      type  hits_this_port  peak_dual_ma
 I    abnormal_device_sender            True          2.19
 A   business_traffic_growth            True          1.80
 E        link_quality_issue           False          0.75
 C         large_file_backup           False         -0.66
 G           broadcast_storm            True          0.61
 H        multicast_flooding            True         -0.46
 B         small_packet_scan           False          0.18
 F load_sensitive_link_issue           False         -0.10
 J     unknown_protocol_scan           False          0.10
 D          queue_congestion           False          0.05

同一條線上 |dual MA| 的日常水準  median 0.41   p90 0.98

CH-02 宣告 permanent level shift upward；平日 09:00-11:00 的實測中位數
  事件之前  0.0858 MB/s   n=48
  事件之後  0.0860 MB/s   n=432


同一條線上 `|dual MA|` 的中位數是 0.41，90 百分位 0.98。拿這兩個數當尺，2.19 與 1.80 明顯站在日常起伏之外，
其餘八列全部落在裡面。A 與 I 正好是這個 port 上僅有的兩次真正的水位移動。

要留意的是 G 和 H 那兩列。它們確實打在 `port-id7430` 上，dual MA 卻只給 0.61 與 -0.46，
跟純粹背景的 C（-0.66）與 E（0.75）分不開。這是 dual MA 的定義決定的。
broadcast storm 改的是封包的組成，dual MA 從定義上只看位元組的水位。
同一份資料上，`broadcast_ratio` 的 robust 分數把 G 頂到 50.0（第 6.4 節那張表），
`traffic_bps` 的 dual MA 給 0.61；兩個數字都對，因為它們問的不是同一件事。
J 那個 0.10 的理由更單純，事件本身在 `port-id7428`，這段時間在這條線上什麼也沒發生。

### Change point 是調查的入口，不是判決

A 那一次的水位移動在 change calendar 裡登記過。CH-02，2026-02-04 08:50 到 11:05，owner 是 app-team，
`expected_signal` 欄寫著 `permanent level shift upward`，原因是新業務系統上線導流到 office VLAN。
所以 dual MA 給出的 1.80 不構成一則告警的理由。偵測器能講的只有「這裡的水位動了」，
動得對不對要去問別的地方，而 change calendar 是最先該問的那個地方。

順著這條線再走一步，還會查到第二件事。CH-02 宣告的是 permanent，資料卻不同意。
事件之後的平日 09:00 到 11:00，`port-id7430` 的中位數是 0.0860 MB/s；事件之前的同一個時段是 0.0858 MB/s。
水位抬高了兩個小時，然後回到原處。
（事件之前只有兩個平日、48 筆樣本，這個比較很薄，方向倒是清楚。）
變更單上寫的預期，跟變更之後量到的事實，是兩件必須分開存放的東西。把前者當成後者，
事後檢討就會停在一個沒有發生過的事實上。

dual MA 抓到的這種東西正式的名字是 change point，它跟 anomaly 不在同一層。下一節把四個名詞一次排開。

## 10. 四個名詞各自指什麼

早上那組分類到目前為止一個名字都沒有喊過，可是十個事件已經把四種形狀演完了。對起來看：

| 名詞 | 判準 | 這份資料裡的例子 |
| --- | --- | --- |
| point anomaly | 單一讀數自己就遠離整條線，不需要任何脈絡 | 事件 D `queue_congestion`。`traffic_bps` 的 robust 峰值 40.4，是第 7 節那張表上流量那一欄最大的一個，四種 baseline 怎麼比都看得見 |
| contextual anomaly | 讀數要放進特定脈絡才算異常，換個脈絡就是常態 | 事件 B `small_packet_scan`。`avg_pkt_bytes` 掉到 127 bytes，對一條專跑小封包的鏈路這是日常，對這個平時 801 bytes 的 port 才是異常，而聚合過的 `traffic_bps` 完全看不見它 |
| collective anomaly | 每一筆單獨看都不離譜，整批合起來才是異常 | 事件 G 與事件 H。五個 port 在同一個 5 分鐘刻度上一起動，橫跨三台設備，但任何一個 port 單獨看，`traffic_bps` 的 robust 峰值絕對值都不到門檻 4（G 落在 1.5 到 3.7，H 落在 1.1 到 2.3） |
| change point | 水位在某個時刻整段移動 | CH-02，change calendar 登記為 `permanent level shift upward`。第 9 節的 dual MA 在 `port-id7430` 上量到 1.80 |

D 那一列有一個保留要說清楚：它連續 14 筆，嚴格講是一串 point anomaly，這份資料裡沒有孤立的單點尖峰。
分類的用處在於它直接決定你該拿什麼去比。point anomaly 配得上任何一種 baseline；
contextual 逼你先把特徵選對；collective 讓 peer baseline 整個失效；
change point 要的則是兩段之間的落差。

### outlier、anomaly、alert、fault

這四個字在會議上經常混著用，代價是討論了半小時才發現大家講的不是同一件事。這份資料四個都有現成的例子。

| 名詞 | 它是什麼 | 例子 |
| --- | --- | --- |
| outlier | 統計上離群，僅此而已，不保證有意義 | 事件 A、B、D、J 的 `broadcast_ratio` 都是 -50.0（截斷），事件 I 是 -47.9。它們離群的原因是 1h MAD 太小 |
| anomaly | 相對於一個講得出來的 baseline 的偏離 | 事件 C `large_file_backup`，seasonal baseline 上 36 筆全部過門檻，最小 9.9 |
| alert | 決定要吵醒一個人 | 事件 C 應該送出 0 則。它在 change calendar 裡是 CH-01，計畫內的備份 |
| fault | 網路上真的壞掉了一個東西 | 事件 E `link_quality_issue`，`errors_pps` 有 8 筆落在 1.713 到 1.917 |

有兩條線索方向相反，值得記住。事件 C 停在 anomaly 這一層：它的分數是整份資料裡最高的一批，
壞掉的東西一個也沒有，正確的處置是一封通知都不要發。
事件 E 反過來，它是 fault，而統計偵測器從頭到尾沒有把它叫成 anomaly。
線路確實有問題，可是第 8 節量到 `error_rate` 的滾動尺度有 99.8% 的時間塌到下限，
偏離在那條線上根本不成立，最後靠 `errors_pps > 0.05` 那條固定規則接住。
從 outlier 走到 alert，中間每一步都要多一個判斷，沒有一步是自動的。

## 11. 三個必做的 GUI 練習

前面都是用看的。這三個要動手，因為「改一個參數、看它怎麼變、再改回來」是唯一能讓參數變成直覺的方法。

### 練習一：把 3 MAD 改成 2 MAD

在 **Recent history, robust: 1h rolling median, +/- 3 MAD** 這個 panel 上停留，
點右上角出現的 panel menu（三個點）> **Edit**。

陰影帶的上下界是 Python 算好之後推上來的兩個欄位 `robust_lo` 與 `robust_hi`，值是中心 ± 3 MAD。
Grafana 這一端拿到的是算好的結果，公式留在 Python 那一側，所以你不能在 UI 裡把 3 改成 2。
你能做的是從已發布的欄位把 2 MAD 反推出來：`robust_center - robust_lo` 就是 3 MAD，
乘上 2/3 就是 2 MAD。

在 panel editor 下方的 query 區把 mode 切到 **Code**，找到 `robust_lo` 那一條 query，改成：

```promql
aiops_python_result{column="robust_center", port_id=~"$port"}
  - on(port_id) (2/3) *
    (aiops_python_result{column="robust_center", port_id=~"$port"}
       - on(port_id) aiops_python_result{column="robust_lo", port_id=~"$port"})
```

`robust_hi` 那一條改成對應的加法版本：

```promql
aiops_python_result{column="robust_center", port_id=~"$port"}
  + on(port_id) (2/3) *
    (aiops_python_result{column="robust_hi", port_id=~"$port"}
       - on(port_id) aiops_python_result{column="robust_center", port_id=~"$port"})
```

比對符號用 `=~` 而不是 `=`，跟這張 dashboard 上其他 panel 一致，這樣 Port dropdown 選 `All` 的時候不會整片空掉。

`on(port_id)` 也是必要的。兩邊的 series 只有 `column` 這個 label 不一樣，
不加向量匹配修飾詞的話 Prometheus 找不到配對，回傳空集合，panel 一樣是空白而且不報錯。
代價是結果只保留 `port_id` 這個 label，series 因此沒有名字可用。
要補回名字得在查詢編輯器的 **Options** 那一列動手：把 **Legend** 從 `Auto` 切成 `Custom`，
右邊才會出現可以自由輸入的欄位，填 `{{port_id}}`。右側面板的 Legend 區只管顯示與排版，改不了 series 的名稱。

**要觀察的是**：帶子變窄之後，原本待在帶內的正常起伏有多少跑到帶外去了。
在 `port-id7431` 的平常時段掃一遍，數一數，然後跟這組實測數字對照：

| 帶寬 | 正常樣本越界比例 | 換算成每天（5 個 port，288 筆/天） | 事件 C 的 36 筆裡越界幾筆 |
| --- | --- | --- | --- |
| 3 MAD | 1.73% | 25 次 | 6 |
| 2 MAD | 5.49% | 79 次 | 6 |

從 3 收到 2，每天多出 54 次越界，而事件 C 一筆都沒有多抓到。
這一組數字只涵蓋這份資料，它示範了收緊門檻最常見的下場：
噪音以正常樣本的密度增加，訊號卻早就被 baseline 吸收掉了，收再緊也叫不回來。
真正該調的是視窗長度或特徵。Lab 02 會把這個取捨量化成一條曲線。

**做完把兩條 query 改回原本的 `robust_lo` 與 `robust_hi`，不要儲存 dashboard。**
如果不小心存了，跑 `python infra/setup_grafana.py` 可以還原（它是 idempotent 的）。

### 練習二：在 Explore 裡把 baseline 用 PromQL 寫一次

Explore 的開法在 Lab 00 走過一次，開好之後先看原始的線：

```promql
aiops_python_result{column="traffic_bps", port_id="port-id7431"}
```

再把 notebook 裡那三種統計量寫成 PromQL：

```promql
avg_over_time(aiops_python_result{column="traffic_bps", port_id="port-id7431"}[1h])
stddev_over_time(aiops_python_result{column="traffic_bps", port_id="port-id7431"}[1h])
quantile_over_time(0.95, aiops_python_result{column="traffic_bps", port_id="port-id7431"}[1d])
```

三條都會有結果，但**它們算的不是 notebook 算的東西**，而且原因正好是這節課的主題。

`[1h]` 是牆上時間的一小時。在 720x 之下，一小時牆上時間等於 720 小時的資料時間，也就是 30 天，
剛好是整份資料。`avg_over_time(...[1h])` 因此是整個月的平均，一個全域常數。
`[1d]` 更誇張，一天牆上時間對應 720 天資料時間，而資料只有 30 天，所以它涵蓋的是 replay 迴圈跑過的全部。

要在 PromQL 裡表達「資料時間的 T 秒」，你得寫 `[T / REPLAY_SPEED_X]`。1 小時資料時間在 720x 之下是 5 秒，
而 scrape interval 就是 5 秒，一個視窗裡只裝得下一個點，`stddev_over_time` 會回傳 0 或空值。
**在 720x 之下，notebook 的 1 小時視窗在 PromQL 裡是表達不出來的。**

把 exporter 改成 `REPLAY_SPEED_X=60` 重起，換算就乾淨了：1 分鐘牆上時間等於 1 小時資料時間，
1 筆樣本佔 5 秒剛好一個 scrape。這時候

```promql
avg_over_time(aiops_python_result{column="traffic_bps", port_id="port-id7431"}[1m])
stddev_over_time(aiops_python_result{column="traffic_bps", port_id="port-id7431"}[1m])
```

視窗裡裝的正好是 12 個點，跟 notebook 的 `rolling(12).mean()` 與 `rolling(12).std()` 是同一批樣本。
兩邊的數字這時候才可以拿來對。

PromQL 沒有內建的 `median_over_time`，要用 `quantile_over_time(0.5, ...)`；
MAD 則沒有辦法用單一個 PromQL 表達式寫出來，因為它需要兩層巢狀的中位數。
這是一個真實的工程限制，也是這門課把 baseline 算在 Python 端、把呈現留給 Grafana 的原因之一。

### 練習三：加一個同時看兩個 port 的 panel

回到 dashboard，右上角 **Edit** > **Add** > **Visualization**。
Datasource 選 Prometheus，query 切到 **Code**，輸入：

```promql
aiops_python_result{column="broadcast_ratio", port_id=~"port-id7427|port-id7429"}
```

Legend 用練習一那個方法填成 `{{port_id}}`。unit 才是右側面板的事：
`Standard options > Unit` 選 `Percent (0.0-1.0)`，選完之後第 6.4 節那些 0.0469 會顯示成 4.69%。
panel 標題可以寫 `Event G - two ports, one instant`。

這兩個 port 分屬 edge-fw-01 與 core-sw-01，是兩台不同的設備。
把 replay 對到事件 G（720x 之下是第 37 分鐘，或用 `FOCUS = ("2026-02-19 11:00", "2026-02-19 13:00")` 配 60x），
你會看到兩條線在**同一個時間刻度**上一起跳起來，一起維持 55 分鐘，一起掉下去。

這張圖就是 peer baseline 失效的直接證據。把它跟第 6.4 節那張表放在一起看：
robust 對兩條線都給 50.0，peer 對兩條線都給 2 以下。同樣的資料，同樣的門檻，結論相反，
差別只在「跟誰比」。

看完不要儲存 dashboard，或者存了之後用 `python infra/setup_grafana.py` 還原。

## 12. 收尾：把表填完

下面最後一個 cell 印出十列空白的表，欄位是 `feature`、`baseline`、`why`。

`feature` 填你會拿哪一個欄位去看這個故障，證據來自第 7 節那張帶號峰值 z 的表。
`baseline` 填 `rolling` / `robust` / `seasonal` / `peer` / `fixed` 其中之一，證據來自 Grafana 上那四個 panel。
兩張表都要看過才填得動。
`why` 用一句話寫清楚另一個選擇為什麼會失敗，這一欄才是這節課真正的產出。
「`traffic_bps` 配 robust」是答案，「因為 B 的封包數與封包大小互相抵銷，乘積不動」才是理解。

填的時候會遇到三種情況，都要能分辨：

第一種是特徵對了，baseline 隨便選都行。事件 D `queue_congestion` 的 `traffic_bps` robust 分數是 40.4，怎麼比都看得見。
第二種要靠 baseline 分勝負：同一條 `traffic_bps` 上的事件 C，rolling 0 筆過門檻，robust 5 筆，seasonal 36 筆。
剩下的第三種由特徵決定，事件 B 與事件 G 就是這樣，特徵選錯的話四種 baseline 全部無效。

事件 C 還有第四層：它在 change calendar 裡登記為 CH-01，是計畫內的備份作業。
seasonal baseline 給它 36 筆滿分，而正確的處置是一封通知都不要發。
分數高不等於該告警。那是下一節 Lab 02 的題目。

填完跟旁邊的人交換。不一致的地方先不要急著說服對方，把它記下來，
Lab 02 會用 event recall、alerts per day、detection delay 這幾個數字來裁決。

In [10]:
catalog = pd.read_csv(PROJECT_ROOT / "data" / "synthetic" / "synthetic_event_catalog.csv")

worksheet = catalog[["event_id", "event_type", "port_id"]].copy()
worksheet["feature"] = ""
worksheet["baseline"] = ""
worksheet["why"] = ""

print("填表用。feature 從第 7 節的欄名選，baseline 從 rolling / robust / seasonal / peer / fixed 選。")
print()
print(worksheet.to_string(index=False))
print()
print("寫完存成自己的檔案，Lab 02 開頭會用到：")
print("    worksheet.to_csv('my_feature_to_fault.csv', index=False)")

填表用。feature 從第 7 節的欄名選，baseline 從 rolling / robust / seasonal / peer / fixed 選。

event_id                event_type     port_id feature baseline why
       A   business_traffic_growth port-id7430                     
       B         small_packet_scan port-id7428                     
       C         large_file_backup port-id7431                     
       D          queue_congestion port-id7427                     
       E        link_quality_issue port-id7429                     
       F load_sensitive_link_issue port-id7427                     
       G           broadcast_storm       MULTI                     
       H        multicast_flooding       MULTI                     
       I    abnormal_device_sender port-id7430                     
       J     unknown_protocol_scan port-id7428                     

寫完存成自己的檔案，Lab 02 開頭會用到：
    worksheet.to_csv('my_feature_to_fault.csv', index=False)
